stages/stage_2/optimize.py

```diff
@@ -193,19 +198,47 @@ def path_optimizer(r, eval):
     """
 
     t1 = match(FIRST_ATOM_PATTERN, r)
-    if t1:
+    if t1 and non_nil(t1["atom"]):
         node = NodePath(t1["atom"].as_int())
         node = node + LEFT
         return r.to(node.as_short_path())
 
     t1 = match(REST_ATOM_PATTERN, r)
-    if t1:
+    if t1 and non_nil(t1["atom"]):
         node = NodePath(t1["atom"].as_int())
         node = node + RIGHT
         return r.to(node.as_short_path())
     return r
 
 
+QUOTE_PATTERN_1 = assemble("(q . 0)")
+
+
+def quote_null_optimizer(r, eval):
+    """
+    This applies the transform `(q . 0)` => `0`
+    """
+    t1 = match(QUOTE_PATTERN_1, r)
+    if t1 is not None:
+        return r.to(0)
+
+    return r
+
+
+APPLY_NULL_PATTERN_1 = assemble("(a 0 . (: . rest))")
+
+
+def apply_null_optimizer(r, eval):
```

Comment:
This one changes semantics, but probably always in a way that benefits the programmer. Oh, it doesn't right now, because the zero opcode will always error. So there's never a meaning change to tossing the rest of the args, unless the program is introspected. Nice optimization.

Reply:
Actually, `0` will evaluate to `0`, which is then passed to `a`, which runs it, and agains evaluate to `0`